File: backend/src/github/contributions.ts
import { graphql } from "@octokit/graphql";

export type ContributionDay = { date: string; count: number; level: 0 | 1 | 2 | 3 | 4 };
export type ContributionCalendar = { totalContributions: number; weeks: { days: ContributionDay[] }[] };

export function levelForCount(count: number): 0 | 1 | 2 | 3 | 4 {
  if (count <= 0) return 0;
  if (count <= 3) return 1;
  if (count <= 6) return 2;
  if (count <= 9) return 3;
  return 4;
}

type RawDay = { date: string; contributionCount: number };
type Raw = { user: { contributionsCollection: { contributionCalendar: {
  totalContributions: number; weeks: { contributionDays: RawDay[] }[];
} } } };

export function mapCalendar(raw: Raw): ContributionCalendar {
  const cal = raw.user.contributionsCollection.contributionCalendar;
  return {
    totalContributions: cal.totalContributions,
    weeks: cal.weeks.map((w) => ({
      days: w.contributionDays.map((d) => ({
        date: d.date, count: d.contributionCount, level: levelForCount(d.contributionCount),
      })),
    })),
  };
}

const QUERY = `query {
  viewer { contributionsCollection { contributionCalendar {
    totalContributions
    weeks { contributionDays { date contributionCount } }
  } } }
}`;

export async function fetchContributions(userToken: string): Promise<ContributionCalendar> {
  const run = graphql.defaults({ headers: { authorization: `token ${userToken}` } });
  // `viewer` returns the authorizing user; reshape to the mapCalendar input form.
  const data = await run<{ viewer: Raw["user"] }>(QUERY);
  return mapCalendar({ user: data.viewer });
}

File: backend-py/app/github/contributions.py
from . import http

QUERY = """query {
  viewer { contributionsCollection { contributionCalendar {
    totalContributions
    weeks { contributionDays { date contributionCount } }
  } } }
}"""


def level_for_count(count: int) -> int:
    if count <= 0:
        return 0
    if count <= 3:
        return 1
    if count <= 6:
        return 2
    if count <= 9:
        return 3
    return 4


def map_calendar(raw: dict) -> dict:
    cal = raw["user"]["contributionsCollection"]["contributionCalendar"]
    return {
        "totalContributions": cal["totalContributions"],
        "weeks": [
            {"days": [
                {"date": d["date"], "count": d["contributionCount"], "level": level_for_count(d["contributionCount"])}
                for d in w["contributionDays"]
            ]}
            for w in cal["weeks"]
        ],
    }


async def fetch_contributions(user_token: str) -> dict:
    data = await http.request_json(
        "POST", f"{http.GITHUB_API}/graphql",
        headers={"Authorization": f"token {user_token}"},
        json={"query": QUERY},
    )
    return map_calendar({"user": data["data"]["viewer"]})
